# TLR-amplicon

**Pipeline to process amplicon sequencing data of Toll-like receptors**

From raw sequencing data to figures

Beware, scripts contain variables specific to this particular project. In order to use it for your purposes you will need to modify it accordingly.

1. Download data and analyse quality by programs fastqc and multiqc
	data are already separated by samples from HEIDELBERG (all TLR amplicons per sample in one file)

2. Trim bad quality data by program [trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic), script **trimatic_TLR.sh** 
	input is defined in the script by forwardsamplelist.txt (list of forward samples in the raw data directory)
	output goes to folder outtrim 

3. Assemble forward and reverse reads by [PEAR](https://sco.h-its.org/exelixis/web/software/pear/doc.html), script **assembler_TLR.sh**
	input both pair of reads that passed previous filtering (run the script in folder outtrim)
	parameters of assemble can be adjusted (length of overlap -v, min size of assembly -n)
	output goes to folder assembled

4. Split individual TLR amplicons by primer sequence by script [**Split_on_Primer.py**](https://github.com/Y-Lammers/Split_on_Primer)
	input is result of assembly in the folder assembled
	forward primers and corresponding amplicons need to be defined in csv file (amplicon name, forward primer sequence)
	some amplicons may end up with no reads, this will produce warning messages.
	output - separated amplicons in fastq files for each sample - goes to folder ampliconsepar
	example of how to run the script in loop for each sample:
	/TLR_mockers/illumina_data/assembled/$ `for f in `ls *fastq`;do ../../../scripts/Split_on_Primer.py -f $f -p ../../ampliconsepar/primers.csv; done`


5. Collapse identical sequences while recording their frequency in each amplicon file by [fastx_collapser](http://hannonlab.cshl.edu/fastx_toolkit/commandline.html), script **collapser_TLR.sh**
	input is result of previous step, amplicon files, amplicon files are defined in samplepas.txt in this file names of non empty amplicon files should be supplied
	output goes to folder collapsed

6. Separate true alleles from errors by frequency threshold, script **selecttruealleles.py** (in-house)
	run this script in the folder with files generated in the previous step
	in each file frequencies of the first and the second most frequent alleles are compared
	if the ratio of second/first allele frequency is higher than 1/3 amplicon is considered heterozygous
	an both sequences are written to the output file. Otherwise individual is considered homozygous
	and only one sequence, the most frequent is written to the output file.
	output files goes to folder ../true/	
	
7. Assemble complete TLR locus from individual overlapping amplicons, script **assemble_amplicons.py** (in-house)
	input definition file: make list of samples:  `ls trueMT* | cut -d. -f 1 | uniq | sort > samplelist.txt`	
	script can be used to assemble homozygous amplicons, while heterozygous amplicons need to be checked in geneious
	and prepared for program PHASE manually. the script just saves the time to solve the homozygots
	assembled homozygous sequences are stored in folder finalassem
	this script is not ideall but it saves some time

8. Prepare phase input file by **prep_phase_TLRX.py** (in-house)
	the script needs two input files, one file with nucleotide sequences (two lines per sample)
	and another file with individual names and populations (format described in the script)
	output can be adjusted by some minor changes and used as PHASE input

9. Parse PHASE results by script **phase_extractor3.py** (in-house)
	the script takes output of PHASE *.out as the first argument and it outputs
	table with summary statistics per populations, fasta file per population, 
	nexus file with all sequences and haplotypes only (can be loaded to popart) and trait file for popart

10. Analyse protein diversity by script **protein_div.py** (in-house)
	input file with protein sequences should be supplied as an argument of this script
	input file can be generated by translating nucleotide sequence to aminoacid sequence in Geneious

11. Plot protein diversity results by rscript **protein_div.R** (in-house)
	many variables should be modified inside the script

12. Calculate number of synonymous and non-synonymous changes by script **snap.analyse.sh**
	this script wraps [snap](https://www.hiv.lanl.gov/repository/aids-db/PROGS/Snap/)


